### ISSUE TITLE
WARNING: eval from module A to B... incremental compilation may be broken

### ISSUE THREAD

**Issue by samoconnor** [JuliaLang/julia#15299]
I have [a macro](https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L798) that converts takes a function body, wraps it in some boilerplate and uploads it to AWS as a Lambda function.

When I use this macro with a function body that includes `eval()` I get `WARNING: eval from module AWSLambda to Foo:`.

The warning seems to be mistaken because in this case the body of the function is not evaluated by the local compiler at all. It is [converted to a string](https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L845), [saved into a ZIP file](https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L684) and later [compiled and executed remotely](https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L624).

Is there a way to selectively suppress this warning?

Perhaps the code that emits the warning should be doing macro expansion first so that it can see that the `eval()` statement is never reached...?

**Comment by vtjnash** [comment]
and it is `eval`'d: https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L826 on the line that prints the warning

**Comment by samoconnor** [comment]
hmmm, perhaps my macro-fu is not strong enough...

On [L826](https://github.com/samoconnor/AWSLambda.jl/blob/master/src/AWSLambda.jl#L826) is:
`body = eval(Expr(:quote, body))`.

The way I read it, the `eval` here is only evaluating the `quote` expression, and the `body` itself does not get evaluated (i.e. the expression `body` is just being wrapped in a `quote` expression).
As I recall, the purpose of this line was to get rid of `LineNumberNode`s so that interpolating `body` back into a string produced cleaner code.

One thing I can say for sure is that the code in `body` is never being locally executed. I have lots of deployed examples where that would be very noticeable!

Have I misunderstood something?

**Comment by vtjnash** [comment]
`eval(Expr(:quote, body))` can execute arbitrary side effects, for example:

```julia
julia> body = quote
              a
              b
              $(Expr(:$,quote println("inject!") end))
              d
              end;

julia> eval(Expr(:quote, body))
inject!
quote  # none, line 2:
    a # none, line 3:
    b # none, line 4:
    nothing # none, line 5:
    d
end
```

it also probably shouldn't be stripping LineNumberNode, that would be a bug. but it looks to me like the line numbers survived intact above.

**Comment by samoconnor** [comment]
I will have to give the injection issue more thought... (Thanks for pointing that out)

Below is an example of what I was trying to achieve with the LineNumberNodes.
Without the `eval(Expr(:quote, ))` wrapper I don't get code that can be pasted into another file and re-parsed.

```julia
macro foo(expr)
    println("# Foo Code:\n$expr")
end

macro bar(expr)
    expr = eval(Expr(:quote, expr))
    println("# Bar Code:\n$expr")
end

@foo function(i)
    while i < 100
        i = i + 1
        println(i)
    end
end

@bar function(i)
    while i < 100
        i = i + 1
        println(i)
    end
end
```

output:
```julia
$ julia test.jl
# Foo Code:
function (i,)
    $(Expr(:line, 11, symbol("/Users/sam/test.jl")))
    while i < 100
        $(Expr(:line, 12, symbol("/Users/sam/test.jl")))
        i = i + 1
        $(Expr(:line, 13, symbol("/Users/sam/test.jl")))
        println(i)
    end
end
# Bar Code:
function (i,) # /Users/sam/test.jl, line 18:
    while i < 100 # /Users/sam/test.jl, line 19:
        i = i + 1 # /Users/sam/test.jl, line 20:
        println(i)
    end
end
```

**Comment by toivoh** [comment]
You can use the following function to get rid of line numbers in an `Expr`:

    clean(ex) = ex
    function clean(ex::Expr)
        args = map(clean, ex.args)
        if ex.head != :block; return Expr(ex.head, args...); end

        args = collect(filter(ex->(!is_linenumber(ex)), args))
        length(args) == 1 ? args[1] : Expr(ex.head, args...)
    end
    is_linenumber(ex) = isa(ex, LineNumberNode) || isexpr(ex, :line)

Might perhaps be generally useful enough to be added to `Base.Meta`? (Under a different name)

I also notice that printing of `:line` Exprs seems to be broken. Obviously it's still needed since line numbers still seem to be represented that way when passed to macros.

**Comment by toivoh** [comment]
Forgot to add `using Base.Meta.isexpr` to the code snippet.

**Comment by samoconnor** [comment]
Nice one @toivoh !
Thank you.

It certainly seems like a good thing to make something like this generally available. It seems like a fairly basic thing to be able to transform a macro's `Expr` argument into a string such that `parse(string(ex)) == ex`.

**Comment by samoconnor** [comment]
Closed See #15309
